pilot/pkg/serviceregistry/kube/cache.go

```diff
@@ -92,14 +92,18 @@ func (pc *PodCache) event(obj interface{}, ev model.Event) error {
 			case v1.PodPending, v1.PodRunning:
 				log.Debugf("Handling update event %s for pod %s in namespace %s -> %v", ev, pod.Name, pod.Namespace, ip)
 				// add to cache if the pod is running or pending
+				pc.rwMu.Lock()
 				pc.keys[ip] = key
+				pc.rwMu.Unlock()
 				if pc.c.EDSUpdater != nil {
 					pc.c.EDSUpdater.WorkloadUpdate(ip, pod.ObjectMeta.Labels, pod.ObjectMeta.Annotations)
 				}
 			default:
 				// delete if the pod switched to other states and is in the cache
-				if pc.keys[ip] == key {
+				if k, _ := pc.getPodKey(ip); k == key {
```

Comment:
doesn't this need a lock?

Reply:
getPodKey has a lock.

Reply:
And in general - we should move code using locks in short getter/setter methods to further reduce the risk.